pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go

```diff
@@ -193,6 +200,227 @@ type HTTPLocalRateLimitSpec struct {
+// for the upstream host.
+type GlobalRateLimitSpec struct {
+	// TCP defines the global rate limiting specification at the network
+	// level. This has the ultimate effect of rate limiting connections
+	// per unit of time that arrive at the upstream host.
+	// +optional
+	TCP *TCPGlobalRateLimitSpec `json:"tcp,omitempty"`
+
+	// HTTP defines the global rate limiting specification for HTTP traffic.
+	// This has the ultimate effect of rate limiting HTTP requests
+	// per unit of time that arrive at the upstream host.
+	// +optional
+	HTTP *HTTPGlobalRateLimitSpec `json:"http,omitempty"`
+}
+
+// TCPGlobalRateLimitSpec defines the global rate limiting specification
+// for TCP connections.
+type TCPGlobalRateLimitSpec struct {
+	// RateLimitService defines the rate limiting service to use
+	// as a global rate limiter.
+	RateLimitService RateLimitServiceSpec `json:"rateLimitService"`
+
+	// Domain defines a container for a set of rate limits.
+	// All domains known to the Ratelimit service must be globally unique.
+	// They serve as a way to have different rate limit configurations that
+	// don't conflict.
+	Domain string `json:"domain"`
+
+	// Descriptors defines the list of rate limit descriptors to use
+	// in the rate limit service request.
+	Descriptors []TCPRateLimitDescriptor `json:"descriptors"`
+
+	// Timeout defines the timeout interval for calls to the rate limit service.
+	// Defaults to 20ms.
+	// +optional
+	Timeout *metav1.Duration `json:"timeout,omitempty"`
+
+	// FailureModeDeny defines whether to allow traffic in case of
+	// communication failure between rate limiting service and the proxy.
+	// Defaults to false.
+	// +optional
+	FailureModeDeny *bool `json:"failureModeDeny,omitempty"`
+}
+
+// RateLimitServiceSpec defines the Rate Limit Service specification.
+type RateLimitServiceSpec struct {
+	// Host defines the hostname of the rate limiting service.
+	Host string `json:"host"`
+
+	// Port defines the port number of the rate limiting service
+	Port uint16 `json:"port"`
+}
+
+// TCPRateLimitDescriptor defines the rate limit descriptor to use
+// in the rate limit service request for TCP connections.
+type TCPRateLimitDescriptor struct {
+	// Entries defines the list of rate limit descriptor entries.
+	Entries []TCPRateLimitDescriptorEntry `json:"entries"`
+}
+
+// TCPRateLimitDescriptorEntry defines the rate limit descriptor entry as a
+// key-value pair to use in the rate limit service request for TCP connections.
+type TCPRateLimitDescriptorEntry struct {
+	// Key defines the key of the descriptor entry.
+	Key string `json:"key"`
+
+	// Value defines the value of the descriptor entry.
+	Value string `json:"value"`
+}
+
+// HTTPGlobalRateLimitSpec defines the global rate limiting specification
+// for HTTP requests.
+type HTTPGlobalRateLimitSpec struct {
+	// RateLimitService defines the rate limiting service to use
+	// as a global rate limiter.
+	RateLimitService RateLimitServiceSpec `json:"rateLimitService"`
```

Comment:
Clarifying question: what's the difference in convention here between a pointer type vs. a value type? I see a mix here (e.g. the fields of GlobalRateLimitSpec)

Reply:
Per API convention, pointer types are used for optional fields.

Reply:
But the bool referenced in my other comment is a pointer, yet it's not optional. I'm guessing it's to deal with the zero value being false?

Reply:
Could you specify which field are you referring to?

Reply:
```go
// Present defines whether the request matches the criteria
// when the header is present. If set to false, header match
// will be performed based on whether the header is absent.

Present *bool `json:"present,omitempty"`
```

Reply:
Added the `+optional` tag to the `Present` field.